[Bryan Tan] iP

fantablack opened the pull request:
# DukePro
> "The important thing to remember, is not to forget"

DukePro frees your mind of having to remember things you need to do. It's,

* text-based
* easy to learn
* ~~FAST~~ SUPER FAST to use

All you need to do is,
1. download it from [here](https://github.com/fantablack/ip/releases/tag/A-Jar)
2. double-click it
3. add your tasks
4. let it manage your tasks for you 😜

And best of all, it is ***100% FREE!***

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [x] Managing Events
- [ ] Reminders (coming soon) 

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```
public static void main(String[] args) {
        Duke kizer = new Duke();
        kizer.run();
}
```

Commit:
Add Gradle support

fantablack committed:
initial commit with level 1

fantablack committed:
level 2 done

fantablack committed:
added task adding

fantablack committed:
added marking of tasks

fantablack committed:
fixed marking and added unmarking using regex

fantablack committed:
Changed input reading to regex

fantablack committed:
use string arrays to capture user input

fantablack committed:
Move bot logic from main() to own object

fantablack committed:
Add Todo functionality

fantablack committed:
Add Deadline functionality

fantablack committed:
Add Event functionality

fantablack committed:
Add basic test scripts

fantablack committed:
Add empty Todo Exception Handling

fantablack committed:
Add invalid command handling

fantablack committed:
Add delete functionality

fantablack committed:
Add invalid delete index exception handling

fantablack committed:
Update main method to close scanner after program exit

fantablack committed:
Remove custom IO Kattio

fantablack committed:
Rename variables to comply with coding convention

fantablack committed:
Add javadoc

fantablack committed:
Change if-else in run() to switch

fantablack committed:
Update test files

fantablack committed:
Update EXPECTED test file

kerbysoh commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,202 @@
+            String command = inputArray[0];
+
+            switch (command) {
+                case "bye":
```
There should be no indentation for "case" clauses here

kerbysoh commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,202 @@
+        System.out.println(greeting);
+        System.out.println(BAR);
+
+        boolean canContinue = true;
```
You can consider not using this variable canContinue, but rather return from the function if the user input for the "bye" is entered

kerbysoh commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,71 @@
+/**
```
All classes should be placed in a package.

kerbysoh commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,202 @@
+    /**
+     * @param curr Task object to be printed
+     */
+    private void printTask(Task curr) {
```
Well done in the naming of all your functions adhering to the coding standard!

kerbysoh reviewed (commented):
Overall, code is clean and well written. Hope you take a look at some minor changes I've suggested! :)

fantablack committed:
Add create blank file if dont exist from loadTasksFromFile()

fantablack committed:
Complete loading task from file to spec

fantablack committed:
Add save to file to spec

fantablack committed:
Add LocalDateTime Conversion for Deadlines

fantablack committed:
Add LocalDateTime Conversion for Events

cwnm commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,50 @@
+     * @param metaInfo deadline information of the task
+     */
+    public Deadline(boolean isDone, String description, String metaInfo) {
+        super(description, "D");
```
metaInfo is a cool variable name

cwnm commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,219 @@
+     * @param curr Task object to be added
+     */
+    private void processNewTask(Task curr) {
+        tasks.add(curr);
```
Nice use of OOP to keep code short

cwnm commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,104 @@
+                String[] item = line.split("\\|");
+                switch (item[0].replaceAll("\\s", "")) {
+                    case "D":
+                        this.tasks.add(new Deadline(item[1].replaceAll("\\s", "").equals("1"), item[2], item[3]));
```
Interesting use of replace for the formatting of data!

cwnm reviewed (commented):
Overall great code!

fantablack committed:
Update test files with new LocalDateTime format

fantablack committed:
Seperate UI functionality into Ui.java

fantablack committed:
Seperate Parser fucntionality into Parser.java

fantablack committed:
Seperate tasks functionality into TaskList.java

fantablack committed:
Divide classes into packages

fantablack committed:
Remove inaccurate exceptions thrown in javadoc headers

fantablack committed:
Revert switch tom traditional switch

fantablack committed:
Add JUnit tests for Parser and DateTimeCustomFormatter

fantablack committed:
Update gitignore to include lib files

fantablack committed:
Fix incorrect create and read location of /data directory

fantablack committed:
Update gitignore to exclude META-INF

fantablack committed:
Add javadoc to all public classes and methods

fantablack committed:
Add Find functionality

fantablack committed:
Merge branch 'add-gradle-support' into branch-A-Gradle

fantablack committed:
Add gradle support

fantablack committed:
Update files to comply with Checkstyle AB3

fantablack committed:
Update test files to comply with Checkstyle

fantablack committed:
Update build.gradle to include JavaFx

fantablack committed:
Add GUI prototype using traditional JavaFx

fantablack committed:
Port GUI from JavaFx to FXML

fantablack committed:
Add Welcome GUI

fantablack committed:
Convert application to GUI

fantablack committed:
Add Assertions to Response and Tasks

fantablack committed:
Add more assertions

TaskList handle*() methods usually involve getting a particular Task from the TaskList to be handled.

These get() operations can sometimes result in null Task objects getting retrieved due to preceding bugs in the TaskList.

Assert statements are now placed at TaskList.get() methods.

Using Java Assertions allows us to add an extra layer of security to ensure that no null Task objects are getting retrieved for handling.

fantablack committed:
Improve Code Quality

There are a few variable Trip-Ups i.e. variables that are pre-assigned a value but then modified without using the original initial values. Complicated expressions also exist.

Trip-Ups are fixed by removing the unnecessary initializations, and complicated expressions have been extracted into a boolean for use in conditionals.

These are all done to improve code quality and readability.

fantablack committed:
Fix extra whitespace when printing TaskList

fantablack committed:
Merge pull request #2 from fantablack/branch-A-Assertions

Merge Branch A Assertions

fantablack committed:
Merge from master

fantablack committed:
Merge pull request #3 from fantablack/branch-A-CodeQuality

Improve Code Quality

fantablack committed:
Add extra bottom padding for DialogBoxes

fantablack committed:
Fix border-radius in DialogBox

fantablack committed:
Add Ui.png

fantablack committed:
Add User Guide

fantablack committed:
Update README.md

Change commands to lowercase.
Cleaned up redundant java programmer try out section.
Clean up structure of feature section.

fantablack committed:
Update README.md

Add mark and unmark commands to usage guide

fantablack committed:
Merge pull request #4 from fantablack/bug-update-ug

Update User Guide

Change commands to lowercase.
Cleaned up redundant java programmer try out section.
Clean up structure of feature section.
Add mark and unmark commands to usage guide.

fantablack committed:
Add deadline date to Deadline usage guide

fantablack committed:
Merge branch 'bug-update-ug'

fantablack committed:
Update MainWindow to close on bye command

fantablack committed:
Update usage guide to include bye command

fantablack committed:
Fix empty deadlines and events being allowed

fantablack committed:
Fix GUI not showing feedback on invalid mark and unmark index

fantablack committed:
Fix GUI freeze for delete on invalid index

fantablack committed:
Update try catch blocks

fantablack committed:
Add TaskList tests

fantablack committed:
Add mark and unmark tests

fantablack committed:
Add tests for handling todo, deadline and event

fantablack committed:
Remove stacktrace prints from Storage.java

fantablack committed:
Fix unhandled exception when tasklist is empty

fantablack committed:
Remove extra log message when creating a new file

fantablack committed:
Add handleDelete() tests

fantablack committed:
Remove unused import

fantablack committed:
Fix empty descriptions allowed in deadlines

fantablack committed:
Fix empty description being allowed for events

fantablack committed:
Improve list task format

fantablack committed:
Add tests to capture more invalid inputs

fantablack committed:
Fix GUI freeze for invalid mark and unmark

fantablack committed:
Fix GUI freeze for empty find

fantablack committed:
Fix task display formatting

fantablack committed:
Rearrange imports order for TaskListTest,java
